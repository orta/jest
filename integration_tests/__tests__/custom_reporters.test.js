/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
'use strict';

const skipOnWindows = require('skipOnWindows');
const {extractSummary} = require('../utils');
const runJest = require('../runJest');

describe('Custom Reporters Integration', () => {
  skipOnWindows.suite();

  test('valid string format for adding reporters', () => {
    const reporterConfig = {
      reporters: ['<rootDir>/reporters/test_reporter.js'],
    };

    const {status} = runJest('custom_reporters', [
      '--config',
      JSON.stringify(reporterConfig),
      'add.test.js',
    ]);

    expect(status).toBe(0);
  });

  test('valid array format for adding reporters', () => {
    const reporterConfig = {
      reporters: [
        [
          '<rootDir>/reporters/test_reporter.js',
          {'Dmitrii Abramov': 'Awesome'},
        ],
      ],
    };

    const {status} = runJest('custom_reporters', [
      '--config',
      JSON.stringify(reporterConfig),
      'add.test.js',
    ]);

    expect(status).toBe(0);
  });

  test('invalid format for adding reporters', () => {
    const reporterConfig = {
      reporters: [[3243242]],
    };

    const {status, stderr} = runJest('custom_reporters', [
      '--config',
      JSON.stringify(reporterConfig),
      'add.test.js',
    ]);

    expect(status).toBe(1);
    expect(stderr).toMatchSnapshot();
  });

  test('default reporters enabled', () => {
    const {stderr, stdout, status} = runJest('custom_reporters', [
      '--config',
      JSON.stringify({
        reporters: ['default', '<rootDir>/reporters/test_reporter.js'],
      }),
      'add.test.js',
    ]);

    const {summary, rest} = extractSummary(stderr);
    const parsedJSON = JSON.parse(stdout);

    expect(status).toBe(0);
    expect(rest).toMatchSnapshot();
    expect(summary).toMatchSnapshot();
    expect(parsedJSON).toMatchSnapshot();
  });

  test('TestReporter with all tests passing', () => {
    const {stdout, status, stderr} = runJest('custom_reporters', [
      'add.test.js',
    ]);

    const parsedJSON = JSON.parse(stdout);

    expect(status).toBe(0);
    expect(stderr.trim()).toBe('');
    expect(parsedJSON).toMatchSnapshot();
  });

  test('TestReporter with all tests failing', () => {
    const {stdout, status, stderr} = runJest('custom_reporters', [
      'add_fail.test.js',
    ]);

    const parsedJSON = JSON.parse(stdout);

    expect(status).toBe(1);
    expect(stderr.trim()).toBe('');
    expect(parsedJSON).toMatchSnapshot();
  });

  test('IncompleteReporter for flexibility', () => {
    const {stderr, stdout, status} = runJest('custom_reporters', [
      '--no-cache',
      '--config',
      JSON.stringify({
        reporters: ['<rootDir>/reporters/incomplete_reporter.js'],
      }),
      'add.test.js',
    ]);

    expect(status).toBe(0);
    expect(stderr.trim()).toBe('');

    expect(stdout).toMatchSnapshot();
  });
});
