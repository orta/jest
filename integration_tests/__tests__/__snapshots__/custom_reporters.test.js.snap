// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Reporters Integration IncompleteReporter for flexibility 1`] = `
"onRunComplete is called
Passed Tests: 1
Failed Tests: 0
Total Tests: 1
"
`;

exports[`Custom Reporters Integration TestReporter with all tests failing 1`] = `
Object {
  "onRunComplete": Object {
    "called": true,
    "numFailedTests": 1,
    "numPassedTests": 0,
    "numTotalTests": 1,
  },
  "onRunStart": Object {
    "called": true,
    "options": "object",
  },
  "onTestResult": Object {
    "called": true,
    "times": 1,
  },
  "onTestStart": Object {
    "called": true,
    "path": false,
  },
  "options": Object {
    "christoph": "pojer",
    "dmitrii": "abramov",
    "hello": "world",
    "maxWorkers": "<<REPLACED>>",
    "pattern": Object {
      "input": "add_fail.test.js",
      "paths": Array [
        "add_fail.test.js",
      ],
      "shouldTreatInputAsPattern": false,
      "testPathPattern": "add_fail.test.js",
    },
    "testPathPattern": "\\"add_fail.test.js\\"",
  },
}
`;

exports[`Custom Reporters Integration TestReporter with all tests passing 1`] = `
Object {
  "onRunComplete": Object {
    "called": true,
    "numFailedTests": 0,
    "numPassedTests": 1,
    "numTotalTests": 1,
  },
  "onRunStart": Object {
    "called": true,
    "options": "object",
  },
  "onTestResult": Object {
    "called": true,
    "times": 1,
  },
  "onTestStart": Object {
    "called": true,
    "path": false,
  },
  "options": Object {
    "christoph": "pojer",
    "dmitrii": "abramov",
    "hello": "world",
    "maxWorkers": "<<REPLACED>>",
    "pattern": Object {
      "input": "add.test.js",
      "paths": Array [
        "add.test.js",
      ],
      "shouldTreatInputAsPattern": false,
      "testPathPattern": "add.test.js",
    },
    "testPathPattern": "\\"add.test.js\\"",
  },
}
`;

exports[`Custom Reporters Integration default reporters enabled 1`] = `
" PASS  __tests__/add.test.js
  Custom Reporters
    ✓ adds ok

"
`;

exports[`Custom Reporters Integration default reporters enabled 2`] = `
"Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        <<REPLACED>>
Ran all test suites matching \\"add.test.js\\".
"
`;

exports[`Custom Reporters Integration default reporters enabled 3`] = `
Object {
  "onRunComplete": Object {
    "called": true,
    "numFailedTests": 0,
    "numPassedTests": 1,
    "numTotalTests": 1,
  },
  "onRunStart": Object {
    "called": true,
    "options": "object",
  },
  "onTestResult": Object {
    "called": true,
    "times": 1,
  },
  "onTestStart": Object {
    "called": true,
    "path": false,
  },
  "options": Object {
    "maxWorkers": "<<REPLACED>>",
    "pattern": Object {
      "input": "add.test.js",
      "paths": Array [
        "add.test.js",
      ],
      "shouldTreatInputAsPattern": false,
      "testPathPattern": "add.test.js",
    },
    "testPathPattern": "\\"add.test.js\\"",
  },
}
`;

exports[`Custom Reporters Integration invalid format for adding reporters 1`] = `
"● Reporter Validation Error:

  Unexpected value for Path at index 0 of reporter at index 0
  Expected:
    string
  Got:
    number
  Reporter configuration:
    [
      3243242
    ]

  Configuration Documentation:
  https://facebook.github.io/jest/docs/configuration.html

"
`;
